Feature/expandablelist

Lee-Yoonjoo opened the pull request:
Added expandable list with 2 variations.

Lee-Yoonjoo committed:
Login function is done

Lee-Yoonjoo committed:
two lines are added on develop branch

Lee-Yoonjoo committed:
Expandalbe List Starts

Lee-Yoonjoo committed:
ExpansionPanel and ExpansionPanelList

Lee-Yoonjoo committed:
Each Item of Expandablelist to seperate widget

Lee-Yoonjoo committed:
Expandable list is done

Lee-Yoonjoo committed:
Expandable list added listview with listview.builder

mkoechig commented on lib/widgets/expandablelist.dart:
```diff
@@ -0,0 +1,51 @@
+        ExpansionTile(
+          title: Text(title),
+          controlAffinity: ListTileControlAffinity.leading,
+          children: <Widget>[
+            ListTile(title: Text('This is tile number 1')),
+            ListTile(title: Text('This is tile number 2')),
+            ListTile(title: Text('This is tile number 3')),
+            ListTile(title: Text('This is tile number 4')),
+            ListTile(title: Text('This is tile number 5')),
+          ],
```
This could be encapsulated into a createMockData function

Lee-Yoonjoo replied:
Completed!

mkoechig commented on lib/main.dart:
```diff
@@ -32,6 +32,7 @@ class MyApp extends StatelessWidget {
+          shrinkWrap: true,
           itemCount: items.length,
           itemBuilder: (context, index) {
             final item = items[index];
```
MyApp should reside in it's own class file

Lee-Yoonjoo replied:
Okay I will create new class for it

mkoechig commented on lib/repository/repo.dart:
```diff
@@ -1,29 +1,27 @@
+      IconListItem('Fit', '5 minutes ago', Icons.fitness_center),
+      IconListItem('Emergency Information', '8 minutes ago', Icons.warning),
+      IconListItem('Maps', '8 minutes ago', Icons.pin_drop),
+      ExpandableListSection('See all'),
```
Better have two different Expandable Items here instead of two children in the section. One Expandable per Expanding view makes more sense.

Lee-Yoonjoo replied:
I should make IconList as Expandable one (but always expanded) right?

mkoechig replied:
At the time of writing you had two expandable views inside the ExpandableListSection. You could instead make two sections with one expandable view each.

Lee-Yoonjoo replied:
I understand. For now, I have one Section with column of ListTiles and one Section with column of ExpansionTiles.  You want me to try make only one for example ListSection contains two Widget List of each of them?

mkoechig commented on lib/widgets/expandablelist.dart:
```diff
@@ -0,0 +1,51 @@
+          children: [
+            SizedBox(
+              height: MediaQuery.of(context).size.height - 100,
+              child: ListView.builder(
+                itemBuilder: (context, index) {
+                  return ListTile(
+                      leading: CircleAvatar(
+                        child: Text('${index + 1}'),
+                      ),
+                      title: Text('Item ${index + 1}'));
+                },
+                itemCount: 5,
+                shrinkWrap: true,
```
This creates a new listview inside a listview, so it's nested scrolling. Especially on mobile we alwys want to avoid having a scroll view inside another scrollview

Lee-Yoonjoo replied:
Ah okay. I had the similar issue before. I will be careful.

mkoechig commented on lib/widgets/expandablelist.dart:
```diff
@@ -0,0 +1,51 @@
+            ListTile(title: Text('This is tile number 5')),
+          ],
+        ),
+        ExpansionTile(
```
When expanded this tile has double its content's height with half of it being blank

Lee-Yoonjoo replied:
I will remove the nested scrolling.

Lee-Yoonjoo committed:
Deleted nested scroll listview on expandableList_section. Create IconItem model and put the static data into Repo. Removed ~ItemSection files

Lee-Yoonjoo committed:
'Seperate myapp to its own class'

Lee-Yoonjoo committed:
Added createMockData fuction on IconList and ExpandableList

Lee-Yoonjoo committed:
Moved Mock Data to Repo class.

mkoechig commented on lib/repository/repo.dart:
```diff
@@ -1,29 +1,36 @@
+      IconItem('Emergency Information', '8 minutes ago', Icons.warning),
+      IconItem('Maps', '8 minutes ago', Icons.pin_drop),
     ];
+
```
With CMD+SHIFT+L you can apply auto-format to a whole file and prevent these whitespace mistakes :)

Lee-Yoonjoo replied:
Oops. I will do that before push

mkoechig commented on lib/widgets/expandablelist_section.dart:
```diff
@@ -0,0 +1,28 @@
+    );
+  }
+
+  List<int> mockData = Repo().getMockData();
```
This is now a public property of the ExpandableListSection and is initialised during construction.
It creates a new Repository, gets a snapshot of that data and saves it for the runtime in the property.
Usually you would not cache repository data this way but instead call the repository's method each time you need the data.
So it would make more sense to have the repository as a property and call it's method inside of the createMockData.

Lee-Yoonjoo replied:
I make "Repo" as a property on IconListSection and I put "getIconListItems" method into "createMockData" method.
initialize with new Repo().

And for the expandableList, I add "Repo"  as instance of constructor. And throw itself from Repo class.

Both look working well but which way is right..? or both are okay?

mkoechig replied:
Let's go for the constructor way to do it for now. Later we will introduce dependency injection :)

Lee-Yoonjoo replied:
Okay! I will do that.

mkoechig commented on lib/widgets/iconlist_section.dart:
```diff
@@ -23,4 +24,22 @@ class IconListHead implements ListItem {
+    );
+  }
+
+  List<IconItem> iconItemList = Repo().getIconListItems();
```
Here same as in ExpandableListSection

mkoechig commented on lib/widgets/myapp.dart:
```diff
@@ -0,0 +1,41 @@
+          title: const Text(title),
+        ),
+
+        body: ListView.builder(
+          shrinkWrap: true,
+          itemCount: listItems.length,
+          itemBuilder: (context, index) {
+            final item = listItems[index];
+
+            return ListTile(
+              visualDensity: const VisualDensity(horizontal: 0, vertical: -4),
+              title:item.buildHead(context),
+              subtitle: item.buildBody(context),
+            );
+          },
+        ),
```
I would make this list building a function to reduce spaghetti code

Lee-Yoonjoo replied:
I created _buildListView method.

Lee-Yoonjoo committed:
Correct some based on feedback.
1. remove whitespaces.
2. on MyApp class, create _buildListView method to reduce spaghetti code.
3. Make Repo as a Property of IconListSection and ExpandableListSection. Not sure about how to initialize it properly.

mkoechig commented on lib/widgets/expandablelist_section.dart:
```diff
@@ -0,0 +1,32 @@
+
+class ExpandableListSection implements ListItem {
+  final String title;
+  Repo repo;
```
Should be private

mkoechig commented on lib/widgets/expandablelist_section.dart:
```diff
@@ -0,0 +1,32 @@
+    );
+  }
+
+  List<Widget> createMockData() {
```
Should be private as well

mkoechig commented on lib/widgets/iconlist_section.dart:
```diff
@@ -1,17 +1,18 @@
-class IconListHead implements ListItem {
+class IconListSection implements ListItem {
   final String headerText;
+  Repo repo = Repo();
```
Private

mkoechig commented on lib/widgets/iconlist_section.dart:
```diff
@@ -23,4 +24,25 @@ class IconListHead implements ListItem {
+    );
+  }
+
+  List<Widget> createMockData() {
```
Private

Lee-Yoonjoo committed:
Added repo on IconListSecotion counstructor.
And merge improved.
This feature will be closed.

Lee-Yoonjoo committed:
Missed to do some corrections.
Made some properties to private.
This feature will be closed.
